feat: fixing CMS article 404s with improved error handling and retry logic

fairlighteth opened the pull request:
## Summary

Fixes an issue where some learn articles return a 404 "Page Not Found" error on first visit but work properly when refreshed. This was happening because the initial data fetch from Strapi CMS sometimes failed to return data fast enough, causing Next.js to render the 404 page.

## Changes

1. Added retry logic with exponential backoff for CMS article fetches
2. Improved error handling in article page component
3. Created reusable utility functions for retry operations
4. Extended cache duration from 5 minutes to 60 minutes for better performance
5. Added a cache invalidation API endpoint

## To Test

1. Visit a learn article with a long slug like `/learn/understanding-automated-market-makers-a-beginner-s-guide-to-de-fi-liquidity`
   - [ ] Page should load successfully on first visit
   - [ ] Metadata and content should be properly displayed
   - [ ] No "Page Not Found" error should appear

2. Test cache invalidation
   - [ ] Set up the `REVALIDATE_SECRET` environment variable in Vercel
   - [ ] Request the revalidation endpoint: `/api/revalidate?secret=YOUR_SECRET&tag=cms-content`
   - [ ] Verify the response confirms revalidation (should show `"revalidated": true`)

## Background

The issue was related to Next.js data fetching and ISR (Incremental Static Regeneration). When a user visited a dynamic article page for the first time (especially one not pre-rendered during build), a race condition occurred:

- The `getArticleBySlug` function tried to fetch data from the CMS
- Sometimes this data wasn't available fast enough during the initial render
- The article didn't exist yet in the cache, so `notFound()` was returned
- On refresh, the page worked because the data had been cached

Our solution implements:
1. Multiple retry attempts with increasing delay between them
2. Better error handling to avoid showing 404 unnecessarily
3. Longer cache times to improve performance
4. A cache invalidation mechanism for content updates

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an API endpoint to support cache revalidation with secure access and tag-based targeting.
  - Added utility functions for retrying article fetch operations with exponential backoff.
  - Added pagination support with unified article-per-page constants and improved URL handling.

- **Bug Fixes**
  - Improved error handling and robustness on article pages, ensuring fallback behavior and error logging if article data cannot be fetched.
  - Added redirects for invalid or out-of-range article pagination requests.

- **Chores**
  - Increased cache revalidation interval for CMS content to reduce unnecessary fetches.
  - Enhanced input validation and error handling for article retrieval operations.
  - Standardized pagination constants and URL paths for article listings.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

elena-zh reviewed (approved):
Tested the 1st part only: no 404 errors on articles opening

alfetopito reviewed (approved):
Works as described. Some suggestions

fairlighteth commented on apps/cow-fi/app/(learn)/learn/[article]/page.tsx:
```diff
@@ -25,72 +27,97 @@ export async function generateMetadata({ params }: Props): Promise<Metadata> {
+      description: description
+        ? stripHtmlTags(description)
+        : plainContent.length > 150
+          ? stripHtmlTags(plainContent.substring(0, 147)) + '...'
```
Addressed with inline comment

fairlighteth commented on apps/cow-fi/services/cms/index.ts:
```diff
@@ -6,7 +6,7 @@ import { toQueryParams } from 'util/queryParams'
 
 const DEFAULT_PAGE_SIZE = 100
-const CMS_CACHE_TIME = 5 * 60 // 5 min
+const CMS_CACHE_TIME = 60 * 60 // 60 minutes
```
The idea is to 'bust' the cache (see PR summary):

Test cache invalidation

> Set up the REVALIDATE_SECRET environment variable in Vercel
Request the revalidation endpoint: /api/revalidate?secret=YOUR_SECRET&tag=cms-content
Verify the response confirms revalidation (should show "revalidated": true)


We do still need to setup the REVALIDATE_SECRET in Vercel

alfetopito replied:
Ok, what's the value that should be set?
Send me over DM and I'll set it up

fairlighteth commented on apps/cow-fi/util/fetchHelpers.ts:
```diff
@@ -0,0 +1,39 @@
+export async function fetchArticleWithRetry(slug: string, retries = 2, delay = 500): Promise<Article | null> {
+  try {
+    const article = await getArticleBySlug(slug)
+    return article
```
Addressed now

fairlighteth commented on apps/cow-fi/util/fetchHelpers.ts:
```diff
@@ -0,0 +1,39 @@
+ * @param delay - Initial delay in milliseconds
+ * @returns Result of the operation
+ */
+export async function withRetry<T>(operation: () => Promise<T>, retries = 2, delay = 500): Promise<T> {
```
Addressed

fairlighteth commented:
In my last commit I fixed the articles pagination and improved constants management:

1. Fixed pagination bug where page 3+ redirected to page 1 while keeping incorrect URL
2. Added bounds checking for page numbers with redirect to /learn/articles
3. Moved ARTICLES_PER_PAGE to shared constants to prevent component desync
4. Standardized pagination URL structure (/articles for page 1, /articles/2 for page 2)

Resolves issue where pagination breaks after page 2 and shows incorrect article counts.

elena-zh commented:
LGTM!
